### DISCUSSION TITLE
Support for complex numbers?

### DISCUSSION THREAD

**Discussion by VarLad** [JuliaPlots/Makie.jl#4162] [Ideas]
Makie.jl doesn't support complex numbers for data, yet?
Is there a way to plot a 100 x 2 data of complex numbers?

**Comment by auberjine** [comment]
I'm not sure if plotting complex numbers actually makes sense, what one usually do is to plot either the real/imaginary parts, or the modulus/argument separately
What would you like to see when plotting complex numbers ?

**Comment by VarLad** [comment]
If I give a single vector of complex numbers as argument, I expect it to be plot on the complex plane(real part vs imaginary part) (with the axes getting displayed properly)
If I give more vectors/matrices as argument, I expect it to plot the modulus(behave like normally giving numbers)

THANKS FOR RESPONDING！

**Comment by auberjine** [comment]
I think that's why complex number are not supported directly : depending on the application people expect different results ^^

In your first case you can do
`scatter(real(complex_array), imag(complex_array))`

And in your second
`heatmap(abs.(complex_matrix))`

**Comment by VarLad** [comment]
Figured that out
Although I think its better if people could do that with scatter(z) or scatter(z1,z2).
More appealing for the package🙂
Thanks anyway！

**Comment by SimonDanisch** [comment]
Well, there isn't a problem, as long as the behavior depends on the type!

**Comment by VarLad** [comment]
why not use function overloading？
function plot(a::Complex)

end
somewhere in the Makie src or something like that
Works right？

**Comment by SimonDanisch** [comment]
Yeah, that's an option:
http://makie.juliaplots.org/dev/recipes.html

**Comment by Uroc327** [comment]
Is there more detailed documentation for `convert_arguments` available?
I had hoped I could implement this somehow using
```julia
convert_arguments(::Type{<:Scatter}, x::Complex) = (real(x), imag(x),)
```
but I wasn't able to get a call like `scatter([1+5im, 2+6im, 3+7im, 4+8im])` to work.

**Comment by SimonDanisch** [comment]
Well, if you want to plot a vector of complex number, you need to overload the conversion for a vector of complex numbers.
This should work:
```julia
function convert_arguments(::Type{<:Scatter}, numbers::AbstractVector{<: Complex})
     return (map(x-> Point2f0(real(x), imag(x)), numbers),)
end
```

**Comment by Uroc327** [comment]
Thanks so much! I assumed this was recursive.

**Comment by Uroc327** [comment]
For further reference, I used 

```julia
convert_arguments(::Type{<:Scatter}, coords::AbstractVector{<:Complex}) = (map(x -> Point2(float(real(x)), float(imag(x))), coords),)
convert_arguments(::Type{<:Scatter}, coords::AbstractVector{<:Complex}, heights::AbstractVector{<:Real}) = (map((x, z) -> Point3(float(real(x)), float(imag(x)), float(z)), coords, heights),)
```

**Comment by lambertbrady** [comment]
I second the request for built-in support for plotting complex numbers. I'm new to Julia, and decided to try it out specifically for its typing system, but have found it extremely difficult to use so far in practice. Ideally, I would be able to pass in an array of complex numbers and have them plotted in the complex plane. For example, I would like to use something like `lines(zeta.(vals))`, where `zeta` is the complex-valued Riemann Zeta function from the SpecialFunctions package. Instead, I've had to deconstruct the complex values every time: `lines(real.(zeta.(vals)), imag.(zeta.(vals)))`. Essentially, a list of complex numbers should be interpreted the same way as `lines(positions)` shown in the documentation.

**Comment by SimonDanisch** [comment]
Maybe someone can turn those converts into a PR? :)

**Comment by VarLad** [comment]
@Uroc327 Can you make the PR？

**Comment by jonathanlilly** [comment]
I'd also like to second the vote for this.  I use complex numbers often because they make bivariate data (e.g. horizontal velocity or displacement) quite simple.  For complicated plots writing real(z),imag(z) many times becomes tedious.  Naturally  when you write a complex number you mean z = x + im y, so it's natural to expect that behavior.  

... And if someone has the patience to explain to a Julia newbie what I need to do to take advantages of the changes Uroc327 made, I would greatly appreciate it.

**Reply by berjine** [reply]
Personally when I want to easily do a special plot I define a function to do it, this way it is easy afterwards to add custom ticks, labels, scalings, colors, etc

could be simply :
```julia
function zplot(data)
    lines(real.(data), imag.(data))
end
```
or with more formatting to prepare for a publication for example :
``` julia
function zplot(data, fname::String)
    f = Figure(size=(800,600))
    ax = CairoMakie.Axis(f[1,1], title="Title", xlabel="Horizontal velocity", ylabel="Vertical velocity",
    xticks = LinearTicks(10), yticks = LinearTicks(10), xgridvisible = false, ygridvisible = false)
    lines!(ax, real.(data), imag.(data))
    save(fname*".svg", f)
    f
end
```
